npm/webpack-dev-server/src/plugin.ts

```diff
@@ -52,85 +53,81 @@ export default class CypressCTOptionsPlugin {
-        } else if (this.errorEmitted) {
-          // compilation succeed but assets haven't emitted to the output yet
-          this.devServerEvents.emit('dev-server:compile:error', null)
-        }
-      },
-    )
-
-    compiler.hooks.afterEmit.tap(
-      'CypressCTOptionsPlugin',
-      (compilation) => {
-        if (!compilation.getStats().hasErrors()) {
-          this.devServerEvents.emit('dev-server:compile:success')
-        }
-      },
-    )
+  private afterCompile = () => {
+    if (!this.compilation) {
+      return
+    }
+
+    const stats = this.compilation.getStats()
+
+    if (stats.hasErrors()) {
+      this.errorEmitted = true
+
+      // webpack 4: string[]
+      // webpack 5: Error[]
+      const errors = stats.toJson().errors as Array<Error | string> | undefined
+
+      if (!errors || !errors.length) {
+        return
+      }
+
+      this.devServerEvents.emit('dev-server:compile:error', normalizeError(errors[0]))
+    } else if (this.errorEmitted) {
+      // compilation succeed but assets haven't emitted to the output yet
+      this.devServerEvents.emit('dev-server:compile:error', null)
+    }
+  }
+
+  private afterEmit = () => {
+    if (!this.compilation?.getStats().hasErrors()) {
+      this.devServerEvents.emit('dev-server:compile:success')
+    }
+  }
+
+  private onSpecsChange = (specs: Cypress.Cypress['spec'][]) => {
+    if (!this.compilation || _.isEqual(specs, this.files)) {
+      return
+    }
+
+    this.files = specs
+    const inputFileSystem = this.compilation.inputFileSystem
+    const utimesSync: UtimesSync = semver.gt('4.0.0', webpack.version) ? inputFileSystem.fileSystem.utimesSync : fs.utimesSync
+
+    utimesSync(path.resolve(__dirname, 'browser.js'), new Date(), new Date())
   }
 
   /**
    *
    * @param compilation webpack 4 `compilation.Compilation`, webpack 5
    *   `Compilation`
    */
-  private plugin = (compilation: Webpack45Compilation) => {
-    this.devServerEvents.on('dev-server:specs:changed', (specs) => {
-      if (_.isEqual(specs, this.files)) return
-
-      this.files = specs
-      const inputFileSystem = compilation.inputFileSystem
-      const utimesSync: UtimesSync = semver.gt('4.0.0', webpack.version) ? inputFileSystem.fileSystem.utimesSync : fs.utimesSync
-
-      utimesSync(path.resolve(__dirname, 'browser.js'), new Date(), new Date())
-    })
+  private addCompilationHooks = (compilation: Webpack45Compilation) => {
+    this.compilation = compilation
 
-    // Webpack 5
     /* istanbul ignore next */
     if ('NormalModule' in webpack) {
-      webpack.NormalModule.getCompilationHooks(compilation).loader.tap(
-        'CypressCTOptionsPlugin',
-        (context) => this.pluginFunc(context as CypressCTWebpackContext),
-      )
+      // Webpack 5
+      const loader = webpack.NormalModule.getCompilationHooks(compilation).loader
 
-      return
+      loader.tap('CypressCTOptionsPlugin', this.addLoaderContext)
+    } else {
+      // Webpack 4
```

Comment:
It'll be nice to drop webpack 4 support 🤔

Reply:
We can't... @brian-mann btw, this is the kind of impact we'll have if we bundle the dev servers in with the binary. CC @lmiller1990

Reply:
Yeah. This is the *user's* version of webpack, not *our* version. Dropping support would mean that user projects on webpack 4 couldn't be component tested at all.

Reply:
We will have this problem either way unless we split them into `webpack-dev-server-4` and `webpack-dev-server-5` @JessicaSachs right? This problem is because we want to support all webpack versions in a single package instead of pinning to the matching webpack version.

Confusingly enough webpack-dev-server@5 isn't a thing, v4 works for both. If WDSv5 comes out with breaking changes, we will be a very weird place.

Reply:
I'm happy enough with the current setup - having a conditional + some typecasts in order to support multiple versions isn't the worst thing. If we have another version of webpack that starts requiring more work can reevaluate, but at the moment I don't think it'd be worth splitting out versions.